File: juno/statistics.py
from collections import defaultdict
from decimal import Decimal
from typing import Callable, NamedTuple, Optional

import numpy as np
import pandas as pd

from juno.math import floor_multiple
from juno.time import DAY_MS
from juno.utils import unpack_symbol

from .trading import TradingSummary

Operator = Callable[[Decimal, Decimal], Decimal]
_SQRT_365 = np.sqrt(365)


class Statistics(NamedTuple):
    total_return: float
    annualized_return: float
    annualized_volatility: float
    annualized_downside_risk: float
    sharpe_ratio: float
    sortino_ratio: float
    cagr: float

    alpha: float = 0.0
    beta: float = 0.0


class AnalysisSummary(NamedTuple):
    performance: pd.Series
    a_returns: pd.Series
    g_returns: pd.Series
    neg_g_returns: pd.Series

    stats: Statistics


def analyse_benchmark(prices: list[Decimal]) -> AnalysisSummary:
    performance = pd.Series([float(p) for p in prices])
    return _calculate_statistics(performance)


def analyse_portfolio(
    benchmark_g_returns: pd.Series,
    asset_prices: dict[str, list[Decimal]],
    trading_summary: TradingSummary,
    interval: int = DAY_MS,
) -> AnalysisSummary:
    assert trading_summary.end is not None

    start = floor_multiple(trading_summary.start, interval)
    end = floor_multiple(trading_summary.end, interval)

    # Validate we have enough data. It may be that trading ended prematurely and we have more price
    # points available than needed.
    num_ticks = (end - start) // interval
    for asset, prices in asset_prices.items():
        if len(prices) < num_ticks:
            raise ValueError(
                f'Expected at least {num_ticks} price points for {asset} but got {len(prices)}'
            )

    trades = _get_trades_from_summary(trading_summary, interval)
    asset_performance = _get_asset_performance(
        trading_summary, start, end, asset_prices, trades, interval
    )
    portfolio_performance = pd.Series(
        [float(sum(v for v in apd.values())) for apd in asset_performance]
    )

    return _calculate_statistics(portfolio_performance, benchmark_g_returns)


def _get_trades_from_summary(
    summary: TradingSummary, interval: int
) -> dict[int, list[tuple[str, Decimal]]]:
    trades: dict[int, list[tuple[str, Decimal]]] = defaultdict(list)
    for pos in summary.get_positions():
        base_asset, quote_asset = unpack_symbol(pos.symbol)
        # Open.
        time = floor_multiple(pos.open_time, interval)
        day_trades = trades[time]
        day_trades.append((quote_asset, -pos.cost))
        day_trades.append((base_asset, +pos.base_gain))
        # Close.
        time = floor_multiple(pos.close_time, interval)
        day_trades = trades[time]
        day_trades.append((base_asset, -pos.base_cost))
        day_trades.append((quote_asset, +pos.gain))
    return trades


def _get_asset_performance(
    summary: TradingSummary,
    start_day: int,
    end_day: int,
    asset_prices: dict[str, list[Decimal]],
    trades: dict[int, list[tuple[str, Decimal]]],
    interval: int,
) -> list[dict[str, Decimal]]:
    asset_holdings: dict[str, Decimal] = defaultdict(lambda: Decimal('0.0'))
    asset_holdings[summary.quote_asset] = summary.quote

    asset_performances: list[dict[str, Decimal]] = []

    asset_performances.append(
        _get_asset_performances_from_holdings(asset_holdings, asset_prices, 0)
    )

    # Offset the open price, hence enumrate starting from 1.
    for price_i, time_day in enumerate(range(start_day, end_day, interval), 1):
        # Update holdings.
        day_trades = trades.get(time_day)
        if day_trades:
            for asset, size in day_trades:
                asset_holdings[asset] = asset_holdings[asset] + size

        # Update asset performance (mark-to-market portfolio).
        asset_performances.append(
            _get_asset_performances_from_holdings(asset_holdings, asset_prices, price_i)
        )

    return asset_performances


def _get_asset_performances_from_holdings(
    asset_holdings: dict[str, Decimal], asset_prices: dict[str, list[Decimal]], price_i: int
) -> dict[str, Decimal]:
    asset_performance = {k: Decimal('0.0') for k in asset_prices.keys()}
    for asset, prices in asset_prices.items():
        asset_performance[asset] = asset_holdings[asset] * prices[price_i]
    return asset_performance


def _calculate_statistics(
    performance: pd.Series, benchmark_g_returns: Optional[pd.Series] = None
) -> AnalysisSummary:
    a_returns = performance.pct_change().dropna()
    g_returns = np.log(a_returns + 1)
    neg_g_returns = g_returns[g_returns < 0].dropna()

    # Compute statistics.
    total_return = performance.iloc[-1] / performance.iloc[0] - 1
    annualized_return = 365 * g_returns.mean()
    annualized_volatility = _SQRT_365 * g_returns.std(ddof=0)
    annualized_downside_risk = _SQRT_365 * neg_g_returns.std(ddof=0)

    sharpe_ratio = (
        annualized_return / annualized_volatility if annualized_volatility else Decimal('0.0')
    )
    sortino_ratio = (
        annualized_return / annualized_downside_risk if annualized_downside_risk
        else Decimal('0.0')
    )
    cagr = (
        (performance.iloc[-1] / performance.iloc[0])
        ** (1 / (performance.size / 365))
    ) - 1

    # If benchmark provided, calculate alpha and beta.
    alpha, beta = 0.0, 0.0
    if benchmark_g_returns is not None:
        covariance_matrix = pd.concat(
            [g_returns, benchmark_g_returns], axis=1
        ).dropna().cov(ddof=0)
        y = covariance_matrix.iloc[1].iloc[1]
        if y != 0:
            x = covariance_matrix.iloc[0].iloc[1]
            beta = x / y
            alpha = annualized_return - (beta * 365 * benchmark_g_returns.mean())

    return AnalysisSummary(
        performance=performance,
        a_returns=a_returns,
        g_returns=g_returns,
        neg_g_returns=neg_g_returns,
        stats=Statistics(
            total_return=total_return,
            annualized_return=annualized_return,
            annualized_volatility=annualized_volatility,
            annualized_downside_risk=annualized_downside_risk,
            sharpe_ratio=sharpe_ratio,
            sortino_ratio=sortino_ratio,
            cagr=cagr,
            alpha=alpha,
            beta=beta
        )
    )

File: juno_rs/src/statistics.rs
use crate::{
    math::{annualized, floor_multiple, mean, std_deviation},
    time::{serialize_interval, serialize_timestamp},
    trading::{Position, TradingSummary},
};
// use ndarray::prelude::*;
// use ndarray_stats::CorrelationExt;
use serde::Serialize;
use std::collections::HashMap;

pub type AnalysisResult = (f64,);

// TODO: Use const fn when `365.0.sqrt()` is supported.
const SQRT_365: f64 = 19.10497317454279908588432590477168560028076171875;

#[derive(Clone, Copy, Debug, PartialEq, Eq, Hash)]
enum Asset {
    Base,
    Quote,
}

pub struct Statistics {
    // performance: Vec<f64>,
    // a_returns: Vec<f64>,
    pub g_returns: Vec<f64>,
    // neg_g_returns: Vec<f64>,

    // total_return: f64,
    pub annualized_return: f64,
    // annualized_volatility: f64,
    // annualized_downside_risk: f64,
    pub sharpe_ratio: f64,
    pub sortino_ratio: f64,
    // cagr: f64,
}

pub fn analyse(
    summary: &TradingSummary,
    base_prices: &[f64],
    quote_prices: Option<&[f64]>,
    // _benchmark_g_returns: &[f64],
    interval: u64,
) -> Statistics {
    let asset_performance = get_asset_performance(summary, base_prices, quote_prices, interval);
    let portfolio_performance = asset_performance
        .iter()
        .map(|d| d.values().sum())
        .collect::<Vec<f64>>();
    calculate_statistics(&portfolio_performance)
    // let portfolio_stats = calculate_statistics(&portfolio_performance);
    // let (alpha, _beta) = calculate_alpha_beta(&benchmark_g_returns, &portfolio_stats);
    // (alpha,)
    // portfolio_stats
}

fn map_period_deltas_from_summary(
    summary: &TradingSummary,
    interval: u64,
) -> HashMap<u64, Vec<(Asset, f64)>> {
    let mut period_deltas = HashMap::new();
    for pos in summary.positions.iter() {
        let (time, cost, base_gain, close_time, base_cost, gain) = match pos {
            Position::Long(pos) => (
                pos.time,
                pos.cost,
                pos.base_gain,
                pos.close_time,
                pos.base_cost,
                pos.gain,
            ),
            Position::Short(pos) => (
                pos.time,
                pos.cost,
                pos.base_gain,
                pos.close_time,
                pos.base_cost,
                pos.gain,
            ),
        };
        // Open.
        let time = floor_multiple(time, interval);
        let deltas = period_deltas
            .entry(time)
            .or_insert_with(Vec::<(Asset, f64)>::new);
        deltas.push((Asset::Quote, -cost));
        deltas.push((Asset::Base, base_gain));
        // Close.
        let time = floor_multiple(close_time, interval);
        let deltas = period_deltas
            .entry(time)
            .or_insert_with(Vec::<(Asset, f64)>::new);
        deltas.push((Asset::Base, -base_cost));
        deltas.push((Asset::Quote, gain));
    }
    period_deltas
}

fn get_asset_performance(
    summary: &TradingSummary,
    base_prices: &[f64],
    quote_prices: Option<&[f64]>,
    interval: u64,
) -> Vec<HashMap<Asset, f64>> {
    let summary_period_deltas = map_period_deltas_from_summary(summary, interval);

    let start = floor_multiple(summary.start, interval);
    let end = floor_multiple(summary.end, interval);
    let length = ((end - start) / interval) as usize;

    let mut asset_holdings = HashMap::new();
    asset_holdings.insert(Asset::Base, 0.0);
    asset_holdings.insert(Asset::Quote, summary.quote);

    let mut period_asset_performances = Vec::with_capacity(length + 1);

    period_asset_performances.push(get_asset_performances_from_holdings(
        &asset_holdings,
        base_prices[0],
        quote_prices.map(|p| p[0]),
    ));

    for i in 0..length {
        let time = start + i as u64 * interval;
        // Update holdings.
        let deltas = summary_period_deltas.get(&time);
        if let Some(deltas) = deltas {
            for (asset, size) in deltas {
                *asset_holdings.entry(*asset).or_insert(0.0) += size;
            }
        }

        // Update asset performance (mark-to-market portfolio).
        let price_i = i + 1; // Offset the open price.
        period_asset_performances.push(get_asset_performances_from_holdings(
            &asset_holdings,
            base_prices[price_i],
            quote_prices.map(|p| p[price_i]),
        ));
    }

    period_asset_performances
}

fn get_asset_performances_from_holdings(
    asset_holdings: &HashMap<Asset, f64>,
    base_price: f64,
    quote_price: Option<f64>,
) -> HashMap<Asset, f64> {
    // Update asset performance (mark-to-market portfolio).
    let mut asset_performances = HashMap::new();
    for asset in [Asset::Base, Asset::Quote].iter() {
        let entry = asset_performances.entry(*asset).or_insert(0.0);
        *entry = match asset {
            Asset::Base => asset_holdings[asset] * base_price,
            Asset::Quote => asset_holdings[asset] * quote_price.unwrap_or(1.0),
        }
    }
    asset_performances
}

fn calculate_statistics(performance: &[f64]) -> Statistics {
    let mut a_returns = Vec::with_capacity(performance.len() - 1);
    for i in 0..a_returns.capacity() {
        a_returns.push(performance[i + 1] / performance[i] - 1.0);
    }

    let g_returns = a_returns
        .iter()
        .map(|v| (v + 1.0).ln())
        .collect::<Vec<f64>>();
    let annualized_return = 365.0 * mean(&g_returns);

    // Sharpe ratio.
    let sharpe_ratio = if annualized_return.is_nan() || annualized_return == 0.0 {
        0.0
    } else {
        let annualized_volatility = SQRT_365 * std_deviation(&g_returns);
        annualized_return / annualized_volatility
    };

    // Sortino ratio.
    let sortino_ratio = if annualized_return.is_nan() || annualized_return == 0.0 {
        0.0
    } else {
        let neg_g_returns = g_returns
            .iter()
            .cloned()
            .filter(|&v| v < 0.0)
            .collect::<Vec<f64>>();
        let annualized_downside_risk = SQRT_365 * std_deviation(&neg_g_returns);
        // If there are no neg returns, sortino ratio becomes infinite. We will consider it to be
        // 0.0 instead because that is usually a bad run anyway.
        if annualized_downside_risk == 0.0 {
            0.0
        } else {
            annualized_return / annualized_downside_risk
        }
    };

    assert!(sharpe_ratio.is_finite());
    assert!(sortino_ratio.is_finite());

    Statistics {
        // a_returns,
        g_returns,
        // neg_g_returns,
        annualized_return,
        sharpe_ratio,
        sortino_ratio,
    }
}

// fn calculate_alpha_beta(benchmark_g_returns: &[f64], portfolio_stats: &Statistics) -> (f64, f64) {
//     assert!(benchmark_g_returns.len() == portfolio_stats.g_returns.len());

//     // TODO: Inefficient making this copy.
//     let mut combined: Vec<f64> = Vec::with_capacity(benchmark_g_returns.len() * 2);
//     combined.extend(portfolio_stats.g_returns.iter());
//     combined.extend(benchmark_g_returns.iter());

//     let matrix = Array::from_shape_vec((2, benchmark_g_returns.len()), combined)
//         .expect("benchmark and portfolio geometric returns matrix");

//     let covariance_matrix = matrix.cov(0.0).expect("covariance matrix");

//     let beta = covariance_matrix[[0, 1]] / covariance_matrix[[1, 1]];
//     let alpha = portfolio_stats.annualized_return - (beta * 365.0 * mean(&benchmark_g_returns));

//     (alpha, beta)
// }

pub fn get_sharpe_ratio(
    summary: &TradingSummary,
    // Prices have one extra price in the beginning which is the opening price of the first candle.
    base_prices: &[f64],
    quote_prices: Option<&[f64]>,
    interval: u64,
) -> f64 {
    let period_deltas = map_period_deltas_from_summary(summary, interval);

    let start = floor_multiple(summary.start, interval);
    let end = floor_multiple(summary.end, interval);
    let length = ((end - start) / interval) as usize;

    let mut base_holding = 0.0;
    let mut quote_holding = summary.quote;

    let mut prev_performance = match quote_prices {
        Some(quote_prices) => quote_holding * quote_prices[0], // 0 is open price.
        None => quote_holding,
    };

    let mut g_returns = Vec::with_capacity(length);
    let mut sum_g_returns = 0.0;

    for (i, time) in (start..end).step_by(interval as usize).enumerate() {
        let deltas = period_deltas.get(&time);
        if let Some(deltas) = deltas {
            for (asset, size) in deltas {
                match asset {
                    Asset::Base => base_holding += size,
                    Asset::Quote => quote_holding += size,
                }
            }
        }
        let price_i = i + 1; // Offset the open price.
        let performance = base_holding * base_prices[price_i]
            + match quote_prices {
                Some(quote_prices) => quote_holding * quote_prices[price_i],
                None => quote_holding,
            };

        let a_return = performance / prev_performance - 1.0;

        let g_return = (a_return + 1.0).ln();
        g_returns.push(g_return);
        sum_g_returns += g_return;

        prev_performance = performance;
    }

    let mean_g_returns = sum_g_returns / length as f64;
    let annualized_return = 365.0 * mean_g_returns;

    let sharpe_ratio = if annualized_return.is_nan() || annualized_return == 0.0 {
        0.0
    } else {
        let variance = g_returns
            .iter()
            .map(|value| {
                let diff = mean_g_returns - value;
                diff * diff
            })
            .sum::<f64>()
            / length as f64;
        let std_dev = variance.sqrt();
        let annualized_volatility = SQRT_365 * std_dev;
        annualized_return / annualized_volatility
    };

    debug_assert!(sharpe_ratio.is_finite());
    sharpe_ratio
}

#[derive(Debug, Serialize)]
pub struct TradingStats {
    #[serde(serialize_with = "serialize_timestamp")]
    pub start: u64,
    #[serde(serialize_with = "serialize_timestamp")]
    pub end: u64,
    #[serde(serialize_with = "serialize_interval")]
    pub duration: u64,
    pub cost: f64,
    pub gain: f64,
    pub profit: f64,
    pub roi: f64,
    pub annualized_roi: f64,
    pub mean_position_profit: f64,
    #[serde(serialize_with = "serialize_interval")]
    pub mean_position_duration: u64,
    // pub drawdowns: Vec<f64>,
    pub max_drawdown: f64,
    pub mean_drawdown: f64,
    pub num_positions: u32,
    pub num_positions_in_profit: u32,
    pub num_positions_in_loss: u32,
    pub sharpe_ratio: f64,
    pub sortino_ratio: f64,
}

impl TradingStats {
    pub fn from_summary(
        summary: &TradingSummary,
        base_prices: &[f64],
        quote_prices: Option<&[f64]>,
        stats_interval: u64,
    ) -> Self {
        let mut quote = summary.quote;
        let mut max_quote = quote;
        let mut profit = 0.0;

        let mut num_positions_in_profit = 0;
        let mut num_positions_in_loss = 0;

        // let mut drawdowns = Vec::with_capacity(self.positions.len());
        let mut max_drawdown = 0.0;
        let mut total_drawdown = 0.0;

        let mut total_position_duration = 0;

        for pos in summary.positions.iter() {
            let (pos_profit, pos_duration) = match pos {
                Position::Long(pos) => (pos.profit, pos.duration),
                Position::Short(pos) => (pos.profit, pos.duration),
            };

            profit += pos_profit;
            total_position_duration += pos_duration;

            if pos_profit >= 0.0 {
                num_positions_in_profit += 1;
            } else {
                num_positions_in_loss += 1;
            }

            quote += pos_profit;
            max_quote = f64::max(max_quote, quote);
            let drawdown = 1.0 - quote / max_quote;
            // drawdowns.push(drawdown);
            total_drawdown += drawdown;
            max_drawdown = f64::max(max_drawdown, drawdown);
        }

        let (mean_position_profit, mean_position_duration, mean_drawdown) =
            if summary.positions.len() > 0 {
                (
                    profit / summary.positions.len() as f64,
                    total_position_duration / summary.positions.len() as u64,
                    total_drawdown / summary.positions.len() as f64,
                )
            } else {
                (0.0, 0, 0.0)
            };

        let duration = summary.end - summary.start;
        let cost = summary.quote;
        let gain = cost + profit;
        let roi = profit / cost;
        let annualized_roi = annualized(duration, roi);

        let stats = analyse(&summary, &base_prices, quote_prices, stats_interval);

        Self {
            start: summary.start,
            end: summary.end,
            duration,
            cost,
            gain,
            profit,
            roi,
            annualized_roi,
            mean_position_profit,
            mean_position_duration,
            // drawdowns,
            max_drawdown,
            mean_drawdown,
            num_positions: summary.positions.len() as u32,
            num_positions_in_profit,
            num_positions_in_loss,
            sharpe_ratio: stats.sharpe_ratio,
            sortino_ratio: stats.sortino_ratio,
        }
    }
}
